File: frontend/src/sparkfun/lib/curve.js
/**
 * spark.fun bonding curve — client mirror of contracts/SparkCurve.sol.
 *
 * The contract is exact integer arithmetic, so this is BigInt too: a float
 * preview would disagree with settlement in the last digits, and the trade
 * panel would promise a number the chain does not deliver. Every function here
 * mirrors the Solidity line for line, including rounding direction.
 *
 *   baseOut  = virtualBase  * quoteIn / (virtualQuote + quoteIn)
 *   quoteOut = virtualQuote * baseIn  / (virtualBase  + baseIn)
 *
 * tests/test_curve_parity.py pins this file, curve.py and the contract together.
 */

const E18 = 10n ** 18n;

export const TOTAL_SUPPLY = 1_000_000_000n * E18;
export const CURVE_SUPPLY = 800_000_000n * E18;
export const LP_SUPPLY = 200_000_000n * E18;
export const BPS = 10_000n;

export const STANDARD_VIRTUAL_BASE = 1_073_000_000n * E18;
/** Mayhem steepens the curve. Floor is CURVE_SUPPLY or the curve has no solution. */
export const MAYHEM_VIRTUAL_BASE = (1_073_000_000n * E18 * 85n) / 100n;

export const STANDARD_CREATOR_FEE_BPS = 100n;
export const MAYHEM_CREATOR_FEE_BPS = 250n;
export const PROTOCOL_FEE_BPS = 50n;
export const WALLET_CAP_BPS = 1_000n; // 10% of the graduation target

export const PAIRS = {
  ETH: {
    symbol: 'ETH',
    label: 'ETH',
    decimals: 18,
    graduationRaise: 12n * E18,
    address: null, // native
    blurb: "the chain's native gas",
  },
  USDC: {
    symbol: 'USDC',
    label: 'USDC',
    decimals: 6,
    graduationRaise: 36_000n * 10n ** 6n,
    address: null, // set from REACT_APP_USDC_ADDRESS at runtime
    blurb: 'stable price, market cap you can read',
  },
};

const ceilDiv = (a, b) => (a + b - 1n) / b;
const big = (v) => (typeof v === 'bigint' ? v : BigInt(v ?? 0));

export function curveParams({ pair = 'ETH', mayhem = false } = {}) {
  const p = PAIRS[pair] || PAIRS.ETH;
  const virtualBase0 = mayhem ? MAYHEM_VIRTUAL_BASE : STANDARD_VIRTUAL_BASE;
  const graduationRaise = p.graduationRaise;
  return {
    pair: p.symbol,
    decimals: p.decimals,
    mayhem,
    virtualBase0,
    // Derived exactly as the constructor does, so both land on the target.
    virtualQuote0: (graduationRaise * (virtualBase0 - CURVE_SUPPLY)) / CURVE_SUPPLY,
    graduationRaise,
    creatorFeeBps: mayhem ? MAYHEM_CREATOR_FEE_BPS : STANDARD_CREATOR_FEE_BPS,
    protocolFeeBps: PROTOCOL_FEE_BPS,
    walletQuoteCap: mayhem ? 0n : (graduationRaise * WALLET_CAP_BPS) / BPS,
  };
}

export const virtualBase = (p, baseSold) => p.virtualBase0 - big(baseSold);
export const virtualQuote = (p, quoteRaised) => p.virtualQuote0 + big(quoteRaised);

/** Spot price in quote per whole token, as a Number for display only. */
export function spotPrice(p, baseSold, quoteRaised) {
  const vb = virtualBase(p, baseSold);
  const vq = virtualQuote(p, quoteRaised);
  if (vb <= 0n) return 0;
  return Number(vq) / Number(vb);
}

export function marketCap(p, baseSold, quoteRaised) {
  return spotPrice(p, baseSold, quoteRaised) * Number(TOTAL_SUPPLY / E18);
}

export function progress(p, quoteRaised) {
  const raised = big(quoteRaised);
  if (raised >= p.graduationRaise) return 1;
  return Number((raised * 10_000n) / p.graduationRaise) / 10_000;
}

export function quoteToGraduate(p, quoteRaised) {
  const raised = big(quoteRaised);
  return raised >= p.graduationRaise ? 0n : p.graduationRaise - raised;
}

/** Mirrors SparkCurve.previewBuy, including the ceil on an oversized final buy. */
export function quoteBuy(p, baseSold, quoteRaised, quoteIn) {
  let amount = big(quoteIn);
  const sold = big(baseSold);
  if (amount <= 0n) return emptyQuote('buy');

  const vb = virtualBase(p, sold);
  const vq = virtualQuote(p, quoteRaised);
  const net = amount - (amount * p.creatorFeeBps) / BPS - (amount * p.protocolFeeBps) / BPS;

  let baseOut = (vb * net) / (vq + net);
  let refund = 0n;

  const remaining = CURVE_SUPPLY - sold;
  if (baseOut > remaining) {
    baseOut = remaining;
    const netNeeded = ceilDiv(baseOut * vq, vb - baseOut);
    let gross = ceilDiv(netNeeded * BPS, BPS - p.creatorFeeBps - p.protocolFeeBps);
    if (gross >= amount) gross = amount;
    refund = amount - gross;
    amount = gross;
  }

  const creatorFee = (amount * p.creatorFeeBps) / BPS;
  const protocolFee = (amount * p.protocolFeeBps) / BPS;
  const netSpent = amount - creatorFee - protocolFee;
  const nextRaised = big(quoteRaised) + netSpent;

  return {
    side: 'buy',
    baseOut,
    quoteIn: amount,
    refund,
    creatorFee,
    protocolFee,
    nextSold: sold + baseOut,
    nextRaised,
    graduates: nextRaised >= p.graduationRaise || sold + baseOut >= CURVE_SUPPLY,
    priceBefore: spotPrice(p, sold, quoteRaised),
    priceAfter: spotPrice(p, sold + baseOut, nextRaised),
  };
}

/** Mirrors SparkCurve.previewSell. */
export function quoteSell(p, baseSold, quoteRaised, baseIn) {
  const sold = big(baseSold);
  let amount = big(baseIn);
  if (amount <= 0n) return emptyQuote('sell');
  if (amount > sold) amount = sold;

  const gross = (virtualQuote(p, quoteRaised) * amount) / (virtualBase(p, sold) + amount);
  const creatorFee = (gross * p.creatorFeeBps) / BPS;
  const protocolFee = (gross * p.protocolFeeBps) / BPS;
  const nextRaised = big(quoteRaised) - gross;

  return {
    side: 'sell',
    baseIn: amount,
    quoteOut: gross - creatorFee - protocolFee,
    grossQuote: gross,
    creatorFee,
    protocolFee,
    nextSold: sold - amount,
    nextRaised,
    graduates: false,
    priceBefore: spotPrice(p, sold, quoteRaised),
    priceAfter: spotPrice(p, sold - amount, nextRaised),
  };
}

/** Price impact as a Number, for the panel's warning line. */
export function priceImpact(q) {
  if (!q.priceBefore) return 0;
  return (q.priceAfter - q.priceBefore) / q.priceBefore;
}

/** Samples the whole curve for the chart. Floats are fine here — it is a shape. */
export function curveSamples(p, quoteRaised, points = 72) {
  const out = [];
  const target = Number(p.graduationRaise);
  for (let i = 0; i <= points; i++) {
    const raised = BigInt(Math.floor((Number(p.graduationRaise) * i) / points));
    // Invert the invariant: baseSold such that vq0 + raised holds.
    const vq = p.virtualQuote0 + raised;
    const vb = (p.virtualBase0 * p.virtualQuote0) / vq;
    const sold = p.virtualBase0 - vb;
    out.push({
      progress: (i / points) * 100,
      cap: (Number(vq) / Number(vb)) * Number(TOTAL_SUPPLY / E18),
      price: Number(vq) / Number(vb),
      reached: Number(raised) <= Number(quoteRaised ?? 0n),
      raisedShare: target ? Number(raised) / target : 0,
      sold,
    });
  }
  return out;
}

function emptyQuote(side) {
  return {
    side,
    baseOut: 0n, baseIn: 0n, quoteIn: 0n, quoteOut: 0n, grossQuote: 0n,
    refund: 0n, creatorFee: 0n, protocolFee: 0n,
    nextSold: 0n, nextRaised: 0n, graduates: false,
    priceBefore: 0, priceAfter: 0,
  };
}

File: backend/sparkfun/curve.py
"""spark.fun bonding curve — settlement source of truth.

Constant-product curve over *virtual* reserves:

    k        = virtual_base * virtual_quote
    base_out = virtual_base  - k / (virtual_quote + d_quote)   (buy)
    quote_out= virtual_quote - k / (virtual_base  + d_base)    (sell)
    price    = virtual_quote / virtual_base

The curve sells CURVE_SUPPLY tokens. Once the raise reaches the pair's
graduation target the curve closes and LP_SUPPLY plus the raised quote seed a
permanently locked Uniswap V3 position.

This module mirrors frontend/src/sparkfun/lib/curve.js. The client copy exists
only for instant quote previews; every number that touches a balance is
computed here. tests/test_curve_parity.py pins the two together.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

TOTAL_SUPPLY = 1_000_000_000.0
CURVE_SUPPLY = 800_000_000.0   # sold along the curve
LP_SUPPLY = 200_000_000.0      # seeded into Uniswap V3 at graduation

FEES: dict[str, dict[str, float]] = {
    "standard": {"creator": 0.010, "protocol": 0.005},
    # Mayhem pays the creator more; the protocol cut is unchanged.
    "mayhem": {"creator": 0.025, "protocol": 0.005},
}

WALLET_CAP_SHARE = 0.02  # share of CURVE_SUPPLY; Mayhem removes the cap

PAIRS: dict[str, dict[str, Any]] = {
    "ETH": {
        "symbol": "ETH",
        "decimals": 18,
        "graduation_raise": 12.0,
        "virtual_base_0": 1_073_000_000.0,
    },
    "USDC": {
        "symbol": "USDC",
        "decimals": 6,
        "graduation_raise": 36_000.0,
        "virtual_base_0": 1_073_000_000.0,
    },
}

# Less virtual base => faster price impact. The factor has a hard floor:
# virtual_base_0 must stay above CURVE_SUPPLY or the curve has no solution
# (1_073_000_000 * f > 800_000_000 => f > 0.7456). 0.85 keeps a safe margin
# while roughly halving the starting market cap and doubling the final one.
MAYHEM_STEEPNESS = 0.85


class CurveError(ValueError):
    """Raised for trades the curve cannot honour."""


@dataclass(frozen=True)
class CurveParams:
    pair: str
    mayhem: bool
    virtual_base_0: float
    virtual_quote_0: float
    graduation_raise: float
    fees: dict[str, float] = field(default_factory=dict)
    wallet_cap: float | None = None


def curve_params(pair: str = "ETH", mayhem: bool = False) -> CurveParams:
    """Immutable curve parameters for a token.

    virtual_quote_0 is derived so the curve reaches exactly `graduation_raise`
    at the moment CURVE_SUPPLY has been sold.
    """
    p = PAIRS.get(pair, PAIRS["ETH"])
    virtual_base_0 = p["virtual_base_0"] * (MAYHEM_STEEPNESS if mayhem else 1.0)
    remaining = virtual_base_0 - CURVE_SUPPLY
    if remaining <= 0:
        raise CurveError("curve misconfigured: virtual_base_0 must exceed CURVE_SUPPLY")
    virtual_quote_0 = (p["graduation_raise"] * remaining) / CURVE_SUPPLY
    return CurveParams(
        pair=p["symbol"],
        mayhem=mayhem,
        virtual_base_0=virtual_base_0,
        virtual_quote_0=virtual_quote_0,
        graduation_raise=p["graduation_raise"],
        fees=FEES["mayhem"] if mayhem else FEES["standard"],
        wallet_cap=None if mayhem else CURVE_SUPPLY * WALLET_CAP_SHARE,
    )


def reserves(params: CurveParams, base_sold: float) -> dict[str, float]:
    virtual_base = params.virtual_base_0 - base_sold
    k = params.virtual_base_0 * params.virtual_quote_0
    virtual_quote = k / virtual_base
    return {
        "virtual_base": virtual_base,
        "virtual_quote": virtual_quote,
        "k": k,
        "raised": virtual_quote - params.virtual_quote_0,
    }


def spot_price(params: CurveParams, base_sold: float) -> float:
    r = reserves(params, base_sold)
    return r["virtual_quote"] / r["virtual_base"]


def market_cap(params: CurveParams, base_sold: float) -> float:
    return spot_price(params, base_sold) * TOTAL_SUPPLY


def progress(params: CurveParams, base_sold: float) -> float:
    return max(0.0, min(1.0, base_sold / CURVE_SUPPLY))


def quote_to_graduate(params: CurveParams, base_sold: float) -> float:
    return max(0.0, params.graduation_raise - reserves(params, base_sold)["raised"])


def quote_buy(params: CurveParams, base_sold: float, quote_in: float) -> dict[str, Any]:
    """Quote a buy. `quote_in` is gross; fees come off the top, so the number
    the user types is the number that leaves their wallet."""
    if quote_in is None or quote_in <= 0:
        return _empty(params, base_sold, "buy")

    creator = params.fees["creator"]
    protocol = params.fees["protocol"]
    net = quote_in * (1.0 - creator - protocol)

    r = reserves(params, base_sold)
    base_out = r["virtual_base"] - r["k"] / (r["virtual_quote"] + net)

    # The curve never sells more than it has left; the remainder is refunded.
    available = CURVE_SUPPLY - base_sold
    refund = 0.0
    if base_out > available:
        base_out = available
        quote_needed = r["k"] / (r["virtual_base"] - base_out) - r["virtual_quote"]
        gross_needed = quote_needed / (1.0 - creator - protocol)
        refund = max(0.0, quote_in - gross_needed)

    spend = quote_in - refund
    next_sold = base_sold + base_out
    before = spot_price(params, base_sold)
    after = spot_price(params, next_sold)
    return {
        "side": "buy",
        "base_out": base_out,
        "quote_in": spend,
        "refund": refund,
        "creator_fee": spend * creator,
        "protocol_fee": spend * protocol,
        "avg_price": (spend / base_out) if base_out > 0 else 0.0,
        "price_before": before,
        "price_after": after,
        "price_impact": _impact(before, after),
        "next_sold": next_sold,
        "graduates": next_sold >= CURVE_SUPPLY - 1e-9,
    }


def quote_sell(params: CurveParams, base_sold: float, base_in: float) -> dict[str, Any]:
    """Quote a sell. `base_in` is tokens sold; fees come off the proceeds."""
    if base_in is None or base_in <= 0:
        return _empty(params, base_sold, "sell")

    amount = min(base_in, base_sold)
    r = reserves(params, base_sold)
    gross = r["virtual_quote"] - r["k"] / (r["virtual_base"] + amount)

    creator = params.fees["creator"]
    protocol = params.fees["protocol"]
    creator_fee = gross * creator
    protocol_fee = gross * protocol
    next_sold = base_sold - amount
    before = spot_price(params, base_sold)
    after = spot_price(params, next_sold)
    return {
        "side": "sell",
        "base_in": amount,
        "quote_out": gross - creator_fee - protocol_fee,
        "gross_quote": gross,
        "creator_fee": creator_fee,
        "protocol_fee": protocol_fee,
        "avg_price": (gross / amount) if amount > 0 else 0.0,
        "price_before": before,
        "price_after": after,
        "price_impact": _impact(before, after),
        "next_sold": next_sold,
        "graduates": False,
    }


def curve_samples(params: CurveParams, base_sold: float, points: int = 64) -> list[dict[str, float]]:
    out = []
    for i in range(points + 1):
        sold = CURVE_SUPPLY * i / points
        out.append({
            "sold": sold,
            "progress": sold / CURVE_SUPPLY,
            "price": spot_price(params, sold),
            "cap": market_cap(params, sold),
            "reached": sold <= base_sold,
        })
    return out


def graduation_plan(params: CurveParams, base_sold: float) -> dict[str, Any]:
    """What the Uniswap V3 position looks like the moment the curve closes."""
    r = reserves(params, base_sold)
    price = spot_price(params, base_sold)
    return {
        "pair": params.pair,
        "quote_liquidity": r["raised"],
        "base_liquidity": LP_SUPPLY,
        "entry_price": price,
        "market_cap": price * TOTAL_SUPPLY,
        "range": "full",
        "locked": True,
    }


def _impact(before: float, after: float) -> float:
    if before <= 0:
        return 0.0
    return (after - before) / before


def _empty(params: CurveParams, base_sold: float, side: str) -> dict[str, Any]:
    price = spot_price(params, base_sold)
    return {
        "side": side,
        "base_out": 0.0, "base_in": 0.0, "quote_in": 0.0, "quote_out": 0.0,
        "gross_quote": 0.0, "refund": 0.0, "creator_fee": 0.0, "protocol_fee": 0.0,
        "avg_price": price, "price_before": price, "price_after": price,
        "price_impact": 0.0, "next_sold": base_sold, "graduates": False,
    }
